### PR TITLE
Setup Updates: SetupShop and GH Action

### PR DESCRIPTION
**Pull request by ric-evans**

### CODE REVIEW

**Comment by ric-evans** [comment]
@dsschult py3.6 is failing because of something in `cryptography`. Have you seen this before?

> ModuleNotFoundError: No module named 'setuptools_rust'

**Comment by blinkdog** [comment]
> @dsschult py3.6 is failing because of something in `cryptography`. Have you seen this before?
> 
> > ModuleNotFoundError: No module named 'setuptools_rust'

Yes, I've seen that error message pop up before. It happens when trying to install the `cryptography` dependency.
Try using `pip` to install it, and it should give you a big block of red text explaining what's going on.
In that big block of red text, there should be a workaround you can use (define a specific environment variable, IIRC) that will prevent the error from happening.
I don't know if applying that workaround causes trouble for installs in later versions of Python. :thinking:

**Comment by dsschult** [comment]
> > @dsschult py3.6 is failing because of something in `cryptography`. Have you seen this before?
> > > ModuleNotFoundError: No module named 'setuptools_rust'
> 
> Yes, I've seen that error message pop up before. It happens when trying to install the `cryptography` dependency.
> Try using `pip` to install it, and it should give you a big block of red text explaining what's going on.
> In that big block of red text, there should be a workaround you can use (define a specific environment variable, IIRC) that will prevent the error from happening.
> I don't know if applying that workaround causes trouble for installs in later versions of Python. thinking

It might just be an issue of `pip` being tool old, and not picking up the whl that's already built.  Have you tried putting in a step for `pip install --upgrade pip` after installing python?

File: setup.py [hunk 1]
```diff
@@ -4,134 +4,37 @@
...
-    url="https://github.com/WIPACrepo/rest-tools",
-    license="https://github.com/WIPACrepo/rest-tools/blob/master/LICENSE",
-    description="REST tools in python - common code for client and server ",
-    long_description=open(os.path.join(HERE, "README.md")).read(),  # include new-lines
-    long_description_content_type="text/markdown",
-    classifiers=sorted(
-        _python_lang_classifiers()
-        + [_development_status()]
-        + [
+    url="https://github.com/WIPACrepo/wipac-telemetry-prototype",
```
Comment (blinkdog) [inline setup.py:26]: Wait a minute... this doesn't seem like the right URI for `rest_tools` :thinking:

File: .github/workflows/try-setup-install.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    outputs:
+      matrix: ${{ steps.py3_versions.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Python Versions
+        id: py3_versions
+        run: |
+          minmin_maxmax=$(grep -P "\(\(\d+, ?\d+\), ?\(\d+, ?\d+\)\)" -oh setup.py | sed 's/[^0-9]/ /g')
+          IFS=', ' read -r -a array <<< "$minmin_maxmax"
+          min_thru_max_series=$(for i in `seq ${array[1]} ${array[3]}`; do printf "3.$i,"; done | rev | cut -c 2- | rev)
```
Comment (blinkdog) [inline .github/workflows/try-setup-install.yml:18]: I think this `cut -c 2-` might only allow 3.X

Where I've tried to specify (3,10) with LTA, it is trying to run setup with Python 3.1 ... :open_mouth:

I can bump LTA down to (3,9), but I thought you'd like a heads-up bug report.

**Review by blinkdog — CHANGES_REQUESTED** [review]
Small issue popped up when I was trying out the new GitHub Action with LTA.

**Comment by blinkdog** [comment]
> > > @dsschult py3.6 is failing because of something in `cryptography`. Have you seen this before?
> > > > ModuleNotFoundError: No module named 'setuptools_rust'
> > 
> > 
> > Yes, I've seen that error message pop up before. It happens when trying to install the `cryptography` dependency.
> > Try using `pip` to install it, and it should give you a big block of red text explaining what's going on.
> > In that big block of red text, there should be a workaround you can use (define a specific environment variable, IIRC) that will prevent the error from happening.
> > I don't know if applying that workaround causes trouble for installs in later versions of Python. thinking
> 
> It might just be an issue of `pip` being too old, and not picking up the whl that's already built. Have you tried putting in a step for `pip install --upgrade pip` after installing python?

After compiling Python 3.6.13 from source and testing it out in a Python 3.6 environment, I can say this is exactly it.
If I try to install `cryptography` with the default pip 18.1, it bombs out with the rust error.
Upgrade to pip 21.1 and it happily installs without any errors.

File: setup.py (hunk 1 above)
Comment (ric-evans) [reply setup.py:26]: d'oh!

File: .github/workflows/try-setup-install.yml (hunk 2 above)
Comment (ric-evans) [reply .github/workflows/try-setup-install.yml]: it was being interpreted at a number at some point. Adding quotes fixed this

Comment (ric-evans) [reply .github/workflows/try-setup-install.yml]: cut removes a trailing comma

**Comment by ric-evans** [comment]
hmmm, it's still bombing

> Python 3.6.13
> Requirement already satisfied: pip in /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages (21.1.2)

**Comment by dsschult** [comment]
> hmmm, it's still bombing
> 
> > Python 3.6.13
> > Requirement already satisfied: pip in /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages (21.1.2)

I'm changing my reasoning.  Because you're doing ```python setup.py install``` it's not actually running `pip`, just distutils.  Maybe a better thing to do would be ```pip install --editable .```.

**Comment by ric-evans** [comment]
> > hmmm, it's still bombing
> > > Python 3.6.13
> > > Requirement already satisfied: pip in /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages (21.1.2)
> 
> I'm changing my reasoning. Because you're doing `python setup.py install` it's not actually running `pip`, just distutils. Maybe a better thing to do would be `pip install --editable .`.

interesting, are there any tradeoffs doing it this way?

**Pull request merged**